### DISCUSSION TITLE
ENV interpolation within actions

### DISCUSSION THREAD

**Discussion by dudo** [community/community#26590] [Actions]
<p>I’m trying to make a generic action to build a docker image and publish to GitHub Packages. I don’t want people futzing with the file, so I want to make use of ENV. Crazy, right?</p>
<pre><code class="lang-auto">on:
  push:
    branches:
      - develop
name: Docker Latest
jobs:
  docker:
    name: Build and push Docker Image
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2 # https://github.com/actions/checkout
      - name: Set ENV
        run: echo "::set-env name=INPUT_REPOSITORY::${GITHUB_REPOSITORY}/${GITHUB_REPOSITORY##*/}"
      - uses: docker/build-push-action@v1 # https://github.com/marketplace/actions/build-and-push-docker-images
        with:
          username: $GITHUB_ACTOR
          password: ${{ secrets.GITHUB_TOKEN }}
          registry: docker.pkg.github.com
          tags: latest
</code></pre>
<p>I also tried defining the ENV at the top level as well, and that actually seemed to give me the same result. <code>${GITHUB_REPOSITORY##*/}</code> doesn’t seem to be working, though the shell says it’s executing within bash.</p>
<pre><code class="lang-auto">GITHUB_REPOSITORY=foo/bar
echo "::set-env name=INPUT_REPOSITORY::${GITHUB_REPOSITORY}/${GITHUB_REPOSITORY##*/}"
::set-env name=INPUT_REPOSITORY::foo/bar/bar
</code></pre>
<p>This works locally. As it stands, github is telling me that my image needs an image name.</p>
<pre><code class="lang-auto">...
Building image [docker.pkg.github.com/foo/bar:latest]
...
Please add an image name to "foo/bar" tag. e.g. "foo/bar/&lt;image_name&gt;"
</code></pre>
<p>I very much am… Any ideas?</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>Please add repository input in your <code>docker/build-push-action@v1</code> step, like:</p>
<pre><code class="lang-auto">jobs:
  docker:
    name: Build and push Docker Image
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2 # https://github.com/actions/checkout
      - name: Set ENV
        run: echo "::set-env name=INPUT_REPOSITORY::${GITHUB_REPOSITORY}/${GITHUB_REPOSITORY##*/}"
      - uses: docker/build-push-action@v1 # https://github.com/marketplace/actions/build-and-push-docker-images
        with:
          username: $GITHUB_ACTOR
          password: ${{ secrets.GITHUB_TOKEN }}
          registry: docker.pkg.github.com
          repository: ${{env.INPUT_REPOSITORY}}
          tags: latest
</code></pre>
<p>Based on my test, the value of <code>INPUT_REPOSITORY</code> set with <code>set-env</code> command will be replaced by <code>empty</code> when run the docker container action without specify <code>repository</code> input variable.  You still need to specify the input variable under <code>with</code> key in the docker contain action. You could refer to an environment variable there.</p>

**Comment by Yanjingzhu** [comment]
<p>According to the action code <a href="https://github.com/docker/github-actions/blob/0f18e2abad9a4ac2963d2516246787375b5ec917/internal/options/tag.go#L31" rel="nofollow noopener">https://github.com/docker/github-actions/blob/0f18e2abad9a4ac2963d2516246787375b5ec917/internal/options/tag.go#L31</a></p>
<pre><code class="lang-auto">func dockerRepo(github GitHub) string {
	if repo := os.Getenv("INPUT_REPOSITORY"); repo != "" {
		return repo
	}
	return strings.ToLower(github.Repository)
}
</code></pre>
<p>When <code>INPUT_REPOSITORY</code> is empty, repo will use github.Repository. In your scenario, it is foo/bar. That the reason why you got error message :</p>
<pre><code class="lang-auto">Please add an image name to "foo/bar" tag. e.g. "foo/bar/&lt;image_name&gt;"
</code></pre>

**Comment by dudo** [comment]
<p>Thanks! <code>repository: ${{env.INPUT_REPOSITORY}}</code> works, but that seems odd. I’m setting the env explicitly. The code you linked even fetches the env. Weird.</p>

**Comment by Yanjingzhu** [comment]
<aside class="quote group-github-partners" data-username="yanjingzhu" data-post="2" data-topic="120196">
<div class="title">
<div class="quote-controls"></div>
 yanjingzhu:</div>
<blockquote>
<p>Based on my test, the value of <code>INPUT_REPOSITORY</code> set with <code>set-env</code> command will be replaced by <code>empty</code> when run the docker container action without specify <code>repository</code> input variable.</p>
</blockquote>
</aside>
<p>Please see the explanation in my previous reply.</p>

**Comment by aspearman** [comment]
<p>This solution no longer works due to changes in how GitHub handles environment variables</p>
